framework/src/main/java/org/checkerframework/common/value/util/Range.java

```diff
@@ -41,96 +42,142 @@
-     * {@code from} is greater than {@code to}, returns {@link #NOTHING}.
+     * Creates a range using BigInteger type bounds.
      *
-     * @param from the lower bound (inclusive)
-     * @param to the upper bound (inclusive)
+     * <p>If the BigInteger range is wider than the full range of the Long class, return EVERYTHING.
+     *
+     * <p>If one of the BigInteger bounds is out of Long's range and {@link #ignoreOverflow} is
+     * false, convert the bounds to Long type in accordance with Java twos-complement overflow
+     * rules, e.g., Long.MAX_VALUE + 1 is converted to Long.MIN_VALUE.
+     *
+     * <p>If one of the BigInteger bounds is out of Long's range and {@link #ignoreOverflow} is
+     * true, convert the bound that is outside Long's range to max/min value of a Long.
+     *
+     * @param bigFrom the lower bound of the BigInteger range
+     * @param bigTo the upper bound of the BigInteger range
+     * @return a range with Long type bounds converted from the BigInteger range
      */
-    private static Range createRangeOrNothing(long from, long to) {
-        if (from <= to) {
-            return new Range(from, to);
+    private static Range create(BigInteger bigFrom, BigInteger bigTo) {
+        if (ignoreOverflow) {
+            bigFrom = bigFrom.max(BIG_LONG_MIN_VALUE);
+            bigTo = bigTo.min(BIG_LONG_MAX_VALUE);
         } else {
-            return NOTHING;
+            BigInteger bigWidth = bigTo.subtract(bigFrom).add(BigInteger.ONE);
+            if (bigWidth.compareTo(BIG_LONG_WIDTH) > 0) {
+                return EVERYTHING;
+            }
         }
+        long longFrom = bigFrom.longValue();
+        long longTo = bigTo.longValue();
+        return createOrElse(longFrom, longTo, EVERYTHING);
     }
 
     /**
      * Returns a range with its bounds specified by two parameters, {@code from} and {@code to}. If
-     * {@code from} is greater than {@code to}, returns {@link #EVERYTHING}.
+     * {@code from} is greater than {@code to}, returns {@link #NOTHING}.
      *
      * @param from the lower bound (inclusive)
      * @param to the upper bound (inclusive)
-     * @return newly-created Range or EVERYTHING
+     * @return newly created Range or EVERYTHING
      */
-    private static Range createRangeOrEverything(long from, long to) {
-        if (from <= to) {
-            return new Range(from, to);
-        } else {
-            return EVERYTHING;
-        }
+    private static Range createOrNothing(long from, long to) {
+        return createOrElse(from, to, NOTHING);
+    }
+
+    /**
+     * Internal factory that handles creation of Range that may be in overflow or underflow as a
+     * result of an internal cast conversion.
+     *
+     * <p>Overflow and underflow here are any violations of {@code from<=to}.
+     *
+     * @param from given from value.
+     * @param to given to value.
+     * @param underflow.
```

Comment:
This parameter name is a bit confusing. The documentation above talks about overflow and underflow, but then there is only an underflow parameter without explanation. Can you add javadoc and use a better name, maybe `alternative`?

Reply:
done for "alternate"